### ISSUE TITLE
Support $convert (aggregation) - along with utility like - $toObjectId

### ISSUE THREAD

**Issue by PranavKumarCSE** [bwaldvogel/mongo-java-server#141]
From MongoDB 4.0 there is a new operator - `$toObjectId`.
Currently its not being supported and results in exception:

Attaching the exception trace:
```
[ERROR] 2020-05-27 21:37:46.712 [mongo-server-worker2] MongoWireProtocolHandler - failed to handle query MongoQuery(header: MessageHeader(request: 25, responseTo: 0), collection: test.$cmd, query: {"aggregate" : "users.user", "pipeline" : [ { "$match" : { "_id" : 5dc19dcbda63c62963725aef } }, { "$project" : { "_id" : 0, "siteIds" : 1 } }, { "$unwind" : "$siteIds" }, { "$project" : { "siteIds" : { "$toObjectId" : "$siteIds" } } }, { "$lookup" : { "from" : "site.metadata", "localField" : "siteIds", "foreignField" : "_id", "as" : "result" } }, { "$project" : { "result._id" : 1, "result.name" : 1 } }, { "$unwind" : "$result" } ], "cursor" : {}}, returnFieldSelector: null)
de.bwaldvogel.mongo.exception.MongoServerError: [Error 168] Unrecognized expression '$toObjectId'
	at de.bwaldvogel.mongo.backend.aggregation.Expression.evaluateDocumentExpression(Expression.java:1502) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.aggregation.Expression.evaluate(Expression.java:1482) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.aggregation.Expression.evaluateDocument(Expression.java:1445) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.aggregation.stage.ProjectStage.projectDocument(ProjectStage.java:63) ~[mongo-java-server-core-1.25.0.jar:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:1.8.0_192]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:1.8.0_192]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:1.8.0_192]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382) ~[?:1.8.0_192]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_192]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_192]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[?:1.8.0_192]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[?:1.8.0_192]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_192]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[?:1.8.0_192]
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270) ~[?:1.8.0_192]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:1.8.0_192]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[?:1.8.0_192]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382) ~[?:1.8.0_192]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_192]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_192]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:1.8.0_192]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_192]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[?:1.8.0_192]
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.runStages(Aggregation.java:158) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.runStages(Aggregation.java:145) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.computeResult(Aggregation.java:183) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.AbstractMongoDatabase.commandAggregate(AbstractMongoDatabase.java:577) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.AbstractMongoDatabase.handleCommand(AbstractMongoDatabase.java:129) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.backend.AbstractMongoBackend.handleCommand(AbstractMongoBackend.java:237) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.wire.MongoDatabaseHandler.handleCommand(MongoDatabaseHandler.java:154) ~[mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.wire.MongoDatabaseHandler.handleQuery(MongoDatabaseHandler.java:91) [mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.wire.MongoDatabaseHandler.channelRead0(MongoDatabaseHandler.java:71) [mongo-java-server-core-1.25.0.jar:?]
	at de.bwaldvogel.mongo.wire.MongoDatabaseHandler.channelRead0(MongoDatabaseHandler.java:36) [mongo-java-server-core-1.25.0.jar:?]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:321) [netty-codec-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:295) [netty-codec-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_192]
```

Sample code usage for `$toObjectId`
```java
MongoCollection<Document> collection = template.getCollection("users.user");
Map<String, String> siteIdNameMap = new HashMap<>();
try {
    List<Document> pipeline = Arrays.asList(
            new Document().append(MATCH_OP, new Document().append("_id", new ObjectId(userId))),
            new Document().append(PROJECT_OP, new Document().append("_id", 0).append(SITEIDS, 1)),
            new Document().append(UNWIND_OP, SITEIDS_REF),
            new Document().append(PROJECT_OP,
                new Document().append(SITEIDS, new Document().append("$toObjectId", SITEIDS_REF))),
            new Document().append("$lookup",
                new Document().append("from", "site.metadata").append("localField", SITEIDS)
                .append("foreignField", "_id").append("as", "result")),
            new Document().append(PROJECT_OP, new Document().append("result._id", 1).append("result.name", 1)),
            new Document().append(UNWIND_OP, "$result"));
    AggregateIterable<Document> itDoc = collection.aggregate(pipeline);
    itDoc.forEach((Block<Document>)document-> {
        Document documentRes = (Document) document.get("result");
        ObjectId id = (ObjectId) documentRes.get("_id");
        siteIdNameMap.put(id.toHexString(), documentRes.getString("name"));
    });
} catch (MongoException dBException) {
    log.error("DBException occurred while aggregating Documents getSiteIdNameMapForUser : ", dBException);
    throw dBException;
}
```

Let me know if you have any questions on this. This addition will really help us. Thanks in advance.

**Comment by PranavKumarCSE** [comment]
The broader support can be done for - https://docs.mongodb.com/manual/reference/operator/aggregation/convert/
This is a newly added - conversion stage.

**Comment by PranavKumarCSE** [comment]
https://docs.mongodb.com/manual/reference/operator/aggregation/convert/

**Comment by bwaldvogel** [comment]
I would appreciate to receive a PR for this feature.

**Comment by bwaldvogel** [comment]
Support for the `$convert` and `$toObjectId` expression is available in version `1.37.0`.
